### ISSUE TITLE
Need help when using torch jit with an thread pool. (how to use at::set_num_threads correctly)

### ISSUE THREAD

**Issue by w1d2s** [pytorch/pytorch#51859]
Hi, I'm trying to using an thread pool with size N to manage N torch::jit::Module instances, and I want assign one thread to each individual torch::jit::Modules. I'm currently wrapping one torch::jit::Module with a wrapper class, and in the constructor I call at::set_num_threads(1) and at::set_num_interop_threads(1), but it seems not behaving as expected (there being only one working thread doing inference at any time, but not N threads). How should I call at::set_num_threads and at::set_num_interop_threads in my program ?   Thanks for attention.

In short, how can I restrict one torch::jit::Module doing inference with only one working thread, while controlling the concurrency of different inferences by an existing thread pool ?

cc @gmagogsfm

**Comment by SplitInfinity** [comment]
As per [this page of the PyTorch documentation] about inter-op and intra-op parallelism, the inter-op parallelism APIs (`set_num_interop_threads`/`get_num_interop_threads`) set the size of the thread pool (shared by all `Modules` in a process) that powers the `torch.jit.fork/wait` APIs that can be used to execute multiple ops from the same model concurrently, and `set_num_threads`/`get_num_threads` are for parallelism within individual operators.

So, in order to achieve your end goal, you need an application level thread pool with N threads. Neither `at::set_num_interop_threads` nor `at::set_num_threads` will help.

**Comment by w1d2s** [comment]
> As per [this page of the PyTorch documentation] about inter-op and intra-op parallelism, the inter-op parallelism APIs (`set_num_interop_threads`/`get_num_interop_threads`) set the size of the thread pool (shared by all `Modules` in a process) that powers the `torch.jit.fork/wait` APIs that can be used to execute multiple ops from the same model concurrently, and `set_num_threads`/`get_num_threads` are for parallelism within individual operators.
> 
> So, in order to achieve your end goal, you need an application level thread pool with N threads. Neither `at::set_num_interop_threads` nor `at::set_num_threads` will help.

Thanks, I do have an application level thread pool with N threads. So according to my understanding, the libtorch maintains another "global thread pool" and all torch::jit::Modules will submit the ops to this "global thread pool" to run, is it right ?

And I have another question: if there is already an application thread pool, what should I do at initialization phase of each working thread ? should I call at::init_num_threads ?  Or I should do nothing, just simply call run_method of torch::jit::Module ?

**Comment by SplitInfinity** [comment]
> Thanks, I do have an application level thread pool with N threads. So according to my understanding, the libtorch maintains another "global thread pool" and all torch::jit::Modules in one process (even if they are instantiated and used in different threads) will submit the ops to this "global thread pool" to run, is it right ?

I think all *forked* ops will be submitted to this global thread pool. Each of your application level threads will serve as "inference threads" shown the diagram [at the top of this page](https://pytorch.org/docs/stable/notes/cpu_threading_torchscript_inference.html).

> And I have another question: if there is already an application thread pool, what should I do at initialization phase of each working thread ? should I call at::init_num_threads ? Or I should do nothing, just simply call run_method of torch::jit::Module ?

I think you only need to call `run_method` on each `torch::jit::Module` in separate threads. If your modules are using `fork/wait` though, you might be limited by the size of the inter-op parallelism thread pool used in TorchScript.

**Comment by w1d2s** [comment]
> > Thanks, I do have an application level thread pool with N threads. So according to my understanding, the libtorch maintains another "global thread pool" and all torch::jit::Modules in one process (even if they are instantiated and used in different threads) will submit the ops to this "global thread pool" to run, is it right ?
> 
> I think all _forked_ ops will be submitted to this global thread pool. Each of your application level threads will serve as "inference threads" shown the diagram [at the top of this page](https://pytorch.org/docs/stable/notes/cpu_threading_torchscript_inference.html).
> 
> > And I have another question: if there is already an application thread pool, what should I do at initialization phase of each working thread ? should I call at::init_num_threads ? Or I should do nothing, just simply call run_method of torch::jit::Module ?
> 
> I think you only need to call `run_method` on each `torch::jit::Module` in separate threads. If your modules are using `fork/wait` though, you might be limited by the size of the inter-op parallelism thread pool used in TorchScript.

Thanks, I do not use fork/wait in my program.  So if I call run_method on each torch::jit::module in separate threads, I can expect that each of my application level thread will serve as "inference thread" in the diagram mentioned above, is that right ?

**Comment by SplitInfinity** [comment]
I think so, yes.

**Comment by w1d2s** [comment]
> I think so, yes.

Thanks for your reply !
